### PR TITLE
Device authentication

### PR DESCRIPTION
**Pull request by ohadbitt**
**Features:**
- Device authentication through msal

### CODE REVIEW

File: ingest/src/test/java/com/microsoft/azure/kusto/ingest/E2ETest.java [hunk 1]
```diff
@@ -290,20 +290,12 @@ void testStreamingIngestFromStream() throws FileNotFoundException {
...
-    private boolean createClients(ConnectionStringBuilder dmCsb, ConnectionStringBuilder engineCsb) {
-        try {
-            IngestClientFactory.createClient(dmCsb);
-            ClientImpl localIngestClient = new ClientImpl(dmCsb);
-//            assertTrue(showDatabasePrincipals(localIngestClient));
-        } catch (URISyntaxException ex) {
-            Assertions.fail("Failed to create ingest client", ex);
-            return false;
-        }
+    private boolean hasAuthentication(ConnectionStringBuilder engineCsb) {
```
Comment (yihezkel) [inline ingest/src/test/java/com/microsoft/azure/kusto/ingest/E2ETest.java:293]: I'd call this method "canCreateClient"

File: ingest/src/test/java/com/microsoft/azure/kusto/ingest/E2ETest.java [hunk 2]
```diff
@@ -39,7 +39,7 @@
     private static final String databaseName = System.getenv("TEST_DATABASE");
     private static final String appId = System.getenv("APP_ID");
     private static final String appKey = System.getenv("APP_KEY");
-    private static final String tenantId = System.getenv("TENANT_ID");
+    private static final String tenantId = System.getenv().getOrDefault("TENANT_ID","microsoft.com");
```
Comment (yihezkel) [inline ingest/src/test/java/com/microsoft/azure/kusto/ingest/E2ETest.java:42]: This is good, but FYI the auth methods can handle a null authorityId:
    if (authorityId == null) {
        authorityId = ORGANIZATION_URI_SUFFIX;
    }

File: data/src/test/java/com/microsoft/azure/kusto/data/ConnectionStringBuilderTest.java [hunk 3]
```diff
@@ -142,7 +142,18 @@ void stripFederatedAuthFromCSB() throws URISyntaxException {
 
         String clusterUrl = "https://service.uri";
         csb = ConnectionStringBuilder.createWithAadApplicationCredentials(clusterUrl, "id", "appKey");
+
+        // The constructor should strip the url from fed=true
         client = new ClientImpl(csb);
         Assertions.assertEquals(clusterUrl, csb.getClusterUrl());
     }
+
+    @Test
```
Comment (yihezkel) [inline data/src/test/java/com/microsoft/azure/kusto/data/ConnectionStringBuilderTest.java:151]: Positive unit test(s)?

File: data/src/main/java/com/microsoft/azure/kusto/data/auth/ConnectionStringBuilder.java [hunk 4]
```diff
@@ -138,6 +144,21 @@ public static ConnectionStringBuilder createWithUserPrompt(String resourceUri, S
         return csb;
     }
 
+    public static ConnectionStringBuilder createWithDeviceAuthentication(String resourceUri) {
```
Comment (yihezkel) [inline data/src/main/java/com/microsoft/azure/kusto/data/auth/ConnectionStringBuilder.java:147]: createWithDeviceCode or createWithDeviceCodeAuthentication (also next method)

Though most of the Java SDK methods don't include the word "Authentication", the .NET and Python SDKs do. So the latter probably makes more sense for the sake of consistency.

File: data/src/main/java/com/microsoft/azure/kusto/data/auth/ConnectionStringBuilder.java [hunk 5]
```diff
@@ -92,6 +94,10 @@ String getAccessToken() {
         return tokenProvider;
     }
 
+    boolean getUseDeviceAuth() {
```
Comment (yihezkel) [inline data/src/main/java/com/microsoft/azure/kusto/data/auth/ConnectionStringBuilder.java:97]: Per naming convention, getters that return a boolean should start with "is". I would go with "isUseDeviceCodeAuth"

File: data/src/main/java/com/microsoft/azure/kusto/data/auth/ConnectionStringBuilder.java [hunk 6]
```diff
@@ -23,6 +23,7 @@
     private String applicationNameForTracing;
     private String accessToken;
     private Callable<String> tokenProvider;
+    private boolean useDeviceAuth;
```
Comment (yihezkel) [inline data/src/main/java/com/microsoft/azure/kusto/data/auth/ConnectionStringBuilder.java:26]: useDeviceCodeAuth would be more accurate

File: ingest/src/test/java/com/microsoft/azure/kusto/ingest/E2ETest.java [hunk 7]
```diff
@@ -313,40 +305,42 @@ private boolean createClients(ConnectionStringBuilder dmCsb, ConnectionStringBui
...
     @Test
     void testCreateWithUserPrompt() {
-        ConnectionStringBuilder dmCsb = ConnectionStringBuilder.createWithUserPrompt(System.getenv("DM_CONNECTION_STRING"), null, System.getenv("USERNAME_HINT"));
         ConnectionStringBuilder engineCsb = ConnectionStringBuilder.createWithUserPrompt(System.getenv("ENGINE_CONNECTION_STRING"), null, System.getenv("USERNAME_HINT"));
-        assertTrue(createClients(dmCsb, engineCsb));
+        assertTrue(hasAuthentication(engineCsb));
     }
 
+//    This is an interactive aproach
+//    @Test
```
Comment (yihezkel) [inline ingest/src/test/java/com/microsoft/azure/kusto/ingest/E2ETest.java:313]: I suggest uncommenting and annotating it as @Ignore("This is an interactive approach. Remove this line to test manually.")

File: ingest/src/test/java/com/microsoft/azure/kusto/ingest/E2ETest.java (hunk 1 above)
Comment (ohadbitt) [reply ingest/src/test/java/com/microsoft/azure/kusto/ingest/E2ETest.java:293]: But you also use it

Comment (ohadbitt) [reply ingest/src/test/java/com/microsoft/azure/kusto/ingest/E2ETest.java:293]: for executing commands

Comment (ohadbitt) [reply ingest/src/test/java/com/microsoft/azure/kusto/ingest/E2ETest.java:293]: and what you want to check is if the authentication works

**Review by yihezkel — APPROVED** [review]

**Pull request merged**